### ISSUE TITLE
Pure Function Convolutions

### ISSUE THREAD

**Issue by gao462** [pyg-team/pytorch_geometric#4592]
### 🚀 The feature, motivation and pitch

In PyTorch, we have `torch.nn.Linear` and `torch.nn.functional.linear(x, w, b)`.
Do we have similar counterpart? For example, `torch_geometric.nn.GCNConv` and `torch_geometric.nn.functional.gcn_conv(x, edge_index, edge_weight, w, b)`?

**Comment by rusty1s** [comment]
Thanks for the issue. Can you say more about the motivation of this and why it might be useful? As far as I can tell, this may involve a big re-structuring of the code base since we are currently uses classes to decompose the functionality of message passing (`message`, `aggregate`, `update`). As such, this is likely not "fixable" by now.

**Comment by gao462** [comment]
The motivation comes from several corner usage requirement in my case.
1. I am transiting from PyTorch convention to JAX convention (pure function) which I believe will be popular in the future;
2. Pure function can exclude most randomness in DL computation (I understand there are some inevitable randomness on GPU due to bottom implementation);
3. PyTorch is also starting JAX-like design (`functorch`) which gives better support on higher-order gradient computation (e.g., Jacobian, Hessian which is fatal in DL) and I need to use that on GNNs.
4. JAX GNN community is quite immature right now, and I would like continue to use PyG.
Thank you.

**Comment by rusty1s** [comment]
Thanks for sharing. I am curious whether the `make_functional(conv)` routine introduced in `torchfunc` works and already fits your needs.
